File: android/src/main/kotlin/com/cheebeez/radio_player/models/Station.kt
package com.cheebeez.radio_player.models

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

data class Station(
    val id: Int, val title: String, val coverUrl: String, val streamUrl: String
) {

    companion object {
        private val gson = Gson()

        fun fromJson(jsonString: String): Station {
            return gson.fromJson(jsonString, Station::class.java)
        }

        fun fromJsonList(items: List<String>): List<Station> {
            return items.map { jsonString -> fromJson(jsonString) }
        }
    }

    fun toJson(): Map<String, Any> {
        val jsonString = gson.toJson(this)
        val mapType = object : TypeToken<Map<String, Any>>() {}.type
        return gson.fromJson(jsonString, mapType)
    }

    fun toJsonEncoded(): String {
        return gson.toJson(this)
    }
}

File: ios/Classes/Station.swift
//
//  Station.swift
//  radio_player
//
//  Created by Dima Kutko on 28.05.2024.
//

import Foundation

class Station: Codable {
    var id: Int
    var title: String
    var coverUrl: String
    var streamUrl: String

    enum CodingKeys: String, CodingKey {
        case id, title, coverUrl, streamUrl
    }

    init(id: Int, title: String, coverUrl: String, streamUrl: String) {
        self.id = id
        self.title = title
        self.coverUrl = coverUrl
        self.streamUrl = streamUrl
    }

    static func fromJson(_ jsonString: String) -> Station? {
        guard let data = jsonString.data(using: .utf8) else { return nil }
        return try? JSONDecoder().decode(Station.self, from: data)
    }

    static func fromJsonList(_ items: [String]) -> [Station] {
        return items.compactMap { fromJson($0) }
    }

    func toJson() -> String? {
        guard let data = try? JSONEncoder().encode(self) else { return nil }
        return String(data: data, encoding: .utf8)
    }

    func toMap() -> [String: Any]? {
        guard let data = try? JSONEncoder().encode(self) else { return nil }
        return (try? JSONSerialization.jsonObject(with: data, options: .allowFragments)) as? [String: Any]
    }
}
